usr.sbin/ctld/ctld.h

```diff
@@ -79,147 +96,426 @@ struct auth_portal {
+	int				pg_discovery_filter = PG_FILTER_UNKNOWN;
+	bool				pg_foreign = false;
+	bool				pg_assigned = false;
+	std::list<portal_up>	        pg_portals;
+	std::unordered_map<std::string, port *> pg_ports;
+	std::string			pg_offload;
+	std::string			pg_redirection;
+	int				pg_dscp = -1;
+	int				pg_pcp = -1;
+
+	uint16_t			pg_tag = 0;
 };
 
-/* Ports created by the kernel.  Perhaps the "p" means "physical" ? */
-struct pport {
-	TAILQ_ENTRY(pport)		pp_next;
-	TAILQ_HEAD(, port)		pp_ports;
-	struct kports			*pp_kports;
-	char				*pp_name;
+typedef std::unique_ptr<portal_group> portal_group_up;
 
-	uint32_t			pp_ctl_port;
+struct port {
+	port(struct target *target);
+	virtual ~port() = default;
+
+	struct target *target() const { return p_target; }
+	virtual struct auth_group *auth_group() const { return nullptr; }
+	virtual struct portal_group *portal_group() const { return nullptr; }
+
+	virtual bool is_dummy() const { return true; }
+
+	virtual void clear_references();
+
+	bool kernel_add();
+	bool kernel_update(const port *oport);
+	bool kernel_remove();
+
+	virtual bool kernel_create_port() = 0;
+	virtual bool kernel_remove_port() = 0;
+
+protected:
+	struct target			*p_target;
+
+	uint32_t			p_ctl_port = 0;
 };
 
-struct port {
-	TAILQ_ENTRY(port)		p_next;
-	TAILQ_ENTRY(port)		p_pgs;
-	TAILQ_ENTRY(port)		p_pps;
-	TAILQ_ENTRY(port)		p_ts;
-	struct conf			*p_conf;
-	char				*p_name;
-	struct auth_group		*p_auth_group;
+struct portal_group_port final : public port {
+	portal_group_port(struct target *target, struct portal_group *pg,
+	    auth_group_sp ag, uint32_t ctl_port);
+	~portal_group_port() override = default;
+
+	struct auth_group *auth_group() const override
+	{ return p_auth_group.get(); }
+	struct portal_group *portal_group() const override
+	{ return p_portal_group; }
+
+	bool is_dummy() const override;
+
+	void clear_references() override;
+
+	bool kernel_create_port() override;
+	bool kernel_remove_port() override;
+
+private:
+	auth_group_sp			p_auth_group;
 	struct portal_group		*p_portal_group;
-	struct pport			*p_pport;
-	struct target			*p_target;
+};
+
+struct ioctl_port final : public port {
+	ioctl_port(struct target *target, int pp, int vp)
+	    : port(target), p_ioctl_pp(pp), p_ioctl_vp(vp) {}
+	~ioctl_port() override = default;
+
+	bool kernel_create_port() override;
+	bool kernel_remove_port() override;
 
-	bool				p_ioctl_port;
+private:
 	int				p_ioctl_pp;
 	int				p_ioctl_vp;
-	uint32_t			p_ctl_port;
+};
+
+struct kernel_port final : public port {
+	kernel_port(struct target *target, struct pport *pp)
+	    : port(target), p_pport(pp) {}
+	~kernel_port() override = default;
+
+	bool kernel_create_port() override;
+	bool kernel_remove_port() override;
+
+private:
+	struct pport			*p_pport;
 };
 
 struct lun {
-	TAILQ_ENTRY(lun)		l_next;
+	lun(struct conf *conf, const char *name);
+
+	const char *name() const { return l_name.c_str(); }
+	const std::string &path() const { return l_path; }
```

Comment:
I think `foo &bar` for references (rather than `foo& bar`) is more consistent with our existing style of `foo *bar` for pointers.